[Merged by Bors] - feat(data/fin): flesh out API for fin

pechersky opened the pull request:
Provide more API for `fin n`. Lemma names chosen to match equivalent lemmas in `nat`. Does not develop docstrings for the lemmas.

New lemmas:
iff lemmas for comparison
`ne_iff_vne`
`eq_mk_iff_coe_eq`
`succ_le_succ_iff`
`succ_lt_succ_iff`

lemmas about explicit numerals
`val_zero'`
`mk_zero`
`mk_one`
`mk_bit0`
`mk_bit1`
`cast_succ_zero`
`succ_zero_eq_one`
`zero_ne_one`
`pred_one`

lemmas about order
`zero_le`
`succ_pos`
`mk_succ_pos`
`one_pos`
`one_lt_succ_succ`
`succ_succ_ne_one`
`pred_mk_succ`
`cast_succ_lt_last`
`cast_succ_lt_succ`
`lt_succ`
`last_pos`
`le_coe_last`

coe lemmas:
`coe_eq_cast_succ`
`coe_succ_eq_succ`
`coe_nat_eq_last`

succ_above API:
`succ_above_below`
`succ_above_zero`
`succ_above_last`
`succ_above_above`
`succ_above_pos`

addition API:
`add_one_pos`
`pred_add_one`

Co-authored by: Yury Kudryashov urkud@ya.ru

---
<!-- put comments you want to keep out of the PR commit here -->

pechersky committed:
feat(data/fin, data/nat/basic): flesh out API for fin

shingtaklam1324 commented on src/data/nat/basic.lean:
```diff
@@ -263,6 +263,11 @@ end
 
+/-- Decreasing then increasing via `pred` then `succ` is idempotent
+only if the term is not 0 -/
+@[simp] lemma pred_succ_ne_zero (n : ℕ) (h : n ≠ 0) : n.pred.succ = n :=
```
```suggestion
@[simp] lemma succ_pred_eq_of_ne_zero (n : ℕ) (h : n ≠ 0) : n.pred.succ = n :=
```

I think the names of lemmas should be what it says when not using projection notation. The similar lemma with `n > 0` instead is called `succ_pred_eq_of_pos`.

pechersky committed:
Use lemma name as if not using proj notation

Co-authored-by: Shing Tak Lam <shingtaklam1324@gmail.com>

pechersky committed:
Remove extra implicit arguments

pechersky committed:
Remove simp, norm_cast attrs for coe_nat_eq_last

This change makes quadratic_reciprocity compile again, likely.
The lemma could probably have  norm_cast attr.

pechersky committed:
Remove added docstrings

pechersky committed:
Remove simp attr from cons_val_zero' (linter)

pechersky committed:
Additional lemmas on pred and succ_above

pechersky committed:
remove simp attr from succ_above_zero

rwbarton commented on src/data/fin.lean:
```diff
@@ -173,27 +183,46 @@ lemma exists_iff {p : fin n → Prop} : (∃ i, p i) ↔ ∃ i h, p ⟨i, h⟩ :
+lemma succ_lt_succ_iff : a.succ < b.succ ↔ a < b :=
+by { simp only [lt_iff_val_lt_val, succ_val], exact ⟨lt_of_succ_lt_succ, succ_lt_succ⟩ }
+
+lemma succ_ne_succ_iff : a.succ ≠ b.succ ↔ a ≠ b :=
```
This feels like one lemma too many--there are loads of injective functions, and proving for all of them that `f a ≠ f b ↔ a ≠ b` is too much repetition, I think.

rwbarton commented on src/data/fin.lean:
```diff
@@ -173,27 +183,46 @@ lemma exists_iff {p : fin n → Prop} : (∃ i, p i) ↔ ∃ i h, p ⟨i, h⟩ :
+lemma succ_succ_ne_one : fin.succ (fin.succ a) ≠ 1 :=
+by { intro h, rw [←succ_zero_eq_one, succ_inj] at h, exact (fin.succ_ne_zero a) h }
+
+lemma one_lt_succ_succ (a : fin (n + 1)) : (1 : fin (n + 3)) < a.succ.succ :=
```
This can be stated with `n` in place of `n + 1`, which is "more general" (if then `n = 0` then the hypotheses are inconsistent, but that's not a problem).

rwbarton commented on src/data/fin.lean:
```diff
@@ -289,13 +318,62 @@ by simp [eq_iff_veq]
 lemma eq_last_of_not_lt {i : fin (n+1)} (h : ¬ i.val < n) : i = last n :=
 le_antisymm (le_last i) (not_lt.1 h)
 
+lemma zero_ne_not_last_add_one (i : fin (n + 1)) (hl : i < fin.last n) : (0 : fin (n + 1)) ≠ i + 1 :=
```
If this is nontrivial to prove, it should probably be stated in the more useful `0 < i + 1` form.

rwbarton commented on src/data/fin.lean:
```diff
@@ -289,13 +318,62 @@ by simp [eq_iff_veq]
+    exact nat.succ_ne_zero _ h.symm }
+end
+
+lemma zero_ne_one : (0 : fin (n + 2)) ≠ 1 := zero_ne_not_last_add_one 0 zero_lt_last
```
Same here: `0 < 1` would be better.

pechersky replied:
There is a `nat.zero_ne_one` lemma in core, so this was emulating that. I wrote the `ne` lemmas to easily provide them to lemmas like `equiv.swap_apply_of_ne_of_ne`. Swapping within `fin n` is often necessary for dealing with rearranging with a sum term. But I can change it, and those lemmas would use `ne_of_lt _`.

rwbarton commented on src/data/fin.lean:
```diff
@@ -289,13 +318,62 @@ by simp [eq_iff_veq]
   cast_succ (fin.succ j) = fin.succ (cast_succ j) :=
 by simp [fin.ext_iff]
 
+lemma cast_succ_ne_succ (i : fin (n + 1)) : i.cast_succ ≠ i.succ :=
```
Again `<` would be better.

rwbarton commented on src/data/fin.lean:
```diff
@@ -289,13 +318,62 @@ by simp [eq_iff_veq]
+  exact coe_val_of_lt (nat.lt.step a.is_lt),
+end
+
+@[simp] lemma coe_succ_eq_succ (a : fin n) : (a.cast_succ + 1) = a.succ :=
```
These parentheses aren't needed. (Also I guess `(a : fin n)` is already declared as a variable, given the previous lemma.)

rwbarton commented on src/data/fin.lean:
```diff
@@ -289,13 +318,62 @@ by simp [eq_iff_veq]
 
+@[simp] lemma cast_succ_zero : cast_succ (0 : fin (n + 1)) = 0 := rfl
+
+@[simp] lemma zero_lt_last : (0 : fin (n + 2)) < last (n + 1) :=
```
Inequalities are not usually simp lemmas, unless they are expected to verify a side condition which comes up in applying another simp lemma. Do you have a use for this one in mind?

pechersky replied:
`zero_lt_last` is used directly below in `zero_ne_one`, but that isn't a simp lemma, so the simp attr is removed here.

rwbarton commented on src/data/fin.lean:
```diff
@@ -289,13 +318,62 @@ by simp [eq_iff_veq]
+  { simp [a.is_lt, eq_iff_veq, add_def, nat.mod_eq_of_lt] }
+end
+
+@[simp] lemma lt_succ : a.cast_succ < a.succ :=
```
Again, does this make sense as a simp lemma?

rwbarton commented on src/data/fin.lean:
```diff
@@ -308,6 +386,16 @@ lemma cast_le_injective {n₁ n₂ : ℕ} (h : n₁ ≤ n₂) : injective (fin.c
 lemma cast_succ_injective (n : ℕ) : injective (@fin.cast_succ n) :=
 cast_le_injective (le_add_right n 1)
 
+@[simp] lemma succ_above_below (p : fin (n + 1)) (i : fin n) (h : i.val < p.val) :
```
I'm skeptical that this and `succ_above_above` should really be `simp` lemmas since I don't expect `simp` to be able to synthesize the proof `h` often.

pechersky replied:
Should `succ_above` be reformulated as:
```
def succ_above (p : fin (n+1)) (i : fin n) : fin (n+1) :=
if i.cast_succ < p then i.cast_succ else i.succ
```
That way we avoid working on the val field (or subtype value, once the other PR is done), and can provide proofs of the inequality in more ways.

rwbarton commented on src/data/fin.lean:
```diff
@@ -308,6 +386,16 @@ lemma cast_le_injective {n₁ n₂ : ℕ} (h : n₁ ≤ n₂) : injective (fin.c
+  p.succ_above i = i.cast_succ :=
+by { rw [fin.succ_above], split_ifs, refl }
+
+lemma succ_above_zero (i : fin n) : succ_above 0 i = i.succ := rfl
```
This looks like a good `simp` lemma.

rwbarton commented on src/data/matrix/notation.lean:
```diff
@@ -78,7 +78,7 @@ section val
 @[simp] lemma cons_val_zero (x : α) (u : fin m → α) : vec_cons x u 0 = x := rfl
 
-@[simp] lemma cons_val_zero' (h : 0 < m.succ) (x : α) (u : fin m → α) :
+lemma cons_val_zero' (h : 0 < m.succ) (x : α) (u : fin m → α) :
```
This looks like a fine `simp` lemma; any particular reason to remove it as such?

pechersky replied:
Linter complained with 
```
#print matrix.cons_val_zero' /- simp can prove this:
  by simp only [fin.mk_zero_eq_zero, matrix.cons_val_zero]
One of the lemmas above could be a duplicate.
```
I can reorder the lemmas in `data.matrix.notation` to make this a simp lemma again.

jcommelin replied:
No, it's fine like this.

rwbarton commented on src/data/nat/basic.lean:
```diff
@@ -263,6 +263,9 @@ end
 @[simp] lemma pred_one_add (n : ℕ) : pred (1 + n) = n :=
 by rw [add_comm, add_one, pred_succ]
 
+@[simp] lemma succ_pred_eq_of_ne_zero (n : ℕ) (h : n ≠ 0) : n.pred.succ = n :=
```
We already have `succ_pred_eq_of_pos`, which you can use in the proof (or might decide makes this lemma irrelevant). Probably should not be a simp lemma.

pechersky replied:
This would not be a simp lemma because it would be difficult to synthesize the `ne` hypothesis, correct?

pechersky committed:
Incorporate edits from rwbarton

pechersky committed:
Additional simp attr removal

pechersky committed:
Specify which zero_lt_one to use

pechersky committed:
Specifty nat.zero_lt_one

pechersky committed:
Simplify succ_pred proof

pechersky committed:
Remove nat succ_pred lemma

pechersky committed:
ne to lt for cast_succ_lt_last

pechersky committed:
pred_one_add lemma

pechersky committed:
fix broken cast_succ_lt_lemmas

pechersky committed:
lemmas on mk for zero and pred

pechersky committed:
fix squeeze_simp into simp only

pechersky committed:
fix use of cast_succ_ne_last to lt

pechersky committed:
make pred_mk_succ have all explicit args

pechersky committed:
add succ_above_pos

pechersky committed:
rename to prod_add_one

pechersky committed:
rename zero_lt lemmas to _pos

pechersky committed:
use fin_zero_elim over elim0

pechersky committed:
undo zeoo typo

pechersky committed:
add succ_above_last

pechersky committed:
simplify succ_above reduction lemmas

jcommelin commented on src/data/fin.lean:
```diff
@@ -87,6 +90,9 @@ instance fin_to_nat (n : ℕ) : has_coe (fin n) nat := ⟨fin.val⟩
 
 lemma mk_val {m n : ℕ} (h : m < n) : (⟨m, h⟩ : fin n).val = m := rfl
 
+lemma eq_mk_iff_val_eq {k : ℕ} {hk : k < n} : a = ⟨k, hk⟩ ↔ a.val = k :=
```
```suggestion
lemma eq_mk_iff_coe_eq {k : ℕ} {hk : k < n} : a = ⟨k, hk⟩ ↔ (a : ℕ) = k :=
```

jcommelin commented on src/data/fin.lean:
```diff
@@ -114,7 +120,11 @@ lemma val_mul {n : ℕ} :  ∀ a b : fin n, (a * b).val = (a.val * b.val) % n
 lemma one_val {n : ℕ} : (1 : fin (n+1)).val = 1 % (n+1) := rfl
 
-@[simp] lemma zero_val (n : ℕ) : (0 : fin (n+1)).val = 0 := rfl
+@[simp] lemma val_zero' (n) : (0 : fin (n+1)).val = 0 := rfl
```
Please add a `coe_zero` version as well, if it doesn't exist yet.

pechersky replied:
There is one several lines above.

jcommelin commented on src/data/fin.lean:
```diff
@@ -173,27 +183,45 @@ lemma exists_iff {p : fin n → Prop} : (∃ i, p i) ↔ ∃ i h, p ⟨i, h⟩ :
 @[simp] lemma succ_inj {a b : fin n} : a.succ = b.succ ↔ a = b :=
 ⟨λh, succ.inj h, λh, by rw h⟩
 
+lemma succ_le_succ_iff : a.succ ≤ b.succ ↔ a ≤ b :=
```
```suggestion
@[simp] lemma succ_le_succ_iff : a.succ ≤ b.succ ↔ a ≤ b :=
```

jcommelin commented on src/data/fin.lean:
```diff
@@ -173,27 +183,45 @@ lemma exists_iff {p : fin n → Prop} : (∃ i, p i) ↔ ∃ i h, p ⟨i, h⟩ :
+lemma succ_le_succ_iff : a.succ ≤ b.succ ↔ a ≤ b :=
+by { simp only [le_iff_val_le_val, succ_val], exact ⟨le_of_succ_le_succ, succ_le_succ⟩ }
+
+lemma succ_lt_succ_iff : a.succ < b.succ ↔ a < b :=
```
```suggestion
@[simp] lemma succ_lt_succ_iff : a.succ < b.succ ↔ a < b :=
```

jcommelin commented on src/data/fin.lean:
```diff
@@ -280,22 +312,68 @@ by cases a; refl
+
+@[simp] lemma pred_one {n : ℕ} : fin.pred (1 : fin (n + 2)) (ne.symm (ne_of_lt one_pos)) = 0 := rfl
+
+lemma pred_add_one (i : fin (n + 2)) (h : i.val < n + 1) :
```
Please use `coe` instead of `.val`

jcommelin commented on src/data/fin.lean:
```diff
@@ -308,6 +386,19 @@ lemma cast_le_injective {n₁ n₂ : ℕ} (h : n₁ ≤ n₂) : injective (fin.c
 lemma cast_succ_injective (n : ℕ) : injective (@fin.cast_succ n) :=
 cast_le_injective (le_add_right n 1)
 
+lemma succ_above_below (p : fin (n + 1)) (i : fin n) (h : i.val < p.val) :
```
Idem

pechersky committed:
incorporate sugg changes

pechersky committed:
fix broken succ_above_pos cases

pechersky committed:
simplify mk lemma names

pechersky committed:
mk bit0, bit1 lemmas

Co-authored by: Yury Kudryashov urkud@ya.ru

pechersky committed:
cast_succ_lt_succ is valid on all fin n

pechersky committed:
add helper zero_ne_one

pechersky committed:
make succ_succ lemmas take explicit args

jcommelin reviewed (approved):
Thanks :tada:
Let's bump lean core to a new version, so that we can do the follow-up refactor (-;

bors merge
